File: c++/Beginner/hangman/hangman.cpp
#include <iostream>
#include <stdlib.h> //srand and rand for random numbers
#include <stdio.h>
#include <time.h> // allows for time

using namespace std;


//This hangman is made using OOP in mind, could be made smaller but i wanted to fiddle around with OOP concepts at beginner level

const int maxGuess = 6;//max guesses
int guesses;//counter for guesses
const string words[5] = {"danger", "computer", "table", "elephant", "uniform"};
const int arrSize = sizeof(words) / sizeof(string);//sizeof array divided by size of string type (otherwise displays size of array in bytes)
string currentWord; //current word in game
string hiddenWord = ""; //replaces currentWord with ____

//generating random number between 1 and 5 - used for choosing random array element, needed to be paired with time

string displayRules()
{
    return "\n\nYou have 6 guesses, with each guess being a letter from A-Z.\nYou must figure out the word by using these guesses as clues. \nYou get one chance to guess the word correctly. If you guess incorrectly you lose.\n\n";
}

string randomizeWord()
{
    srand (time(NULL)); //this allows for random integers https://www.cplusplus.com/reference/cstdlib/srand/
    return currentWord = words[rand()%arrSize]; //this is selecting a new hangman word by assigning rand()%arrSize as seed number to array index
}

void guessWord()
{
    string guessedWord;
    cout << "\n\nType the word you're trying to guess.\n\n";
    cin >> guessedWord;
    if (guessedWord == currentWord)
    {
        cout << "\n\nCongrats, you guessed correctly. YOU WIN!\n\n";
    }
    else
    {
        cout << "\n\nSorry, you guessed incorrectly. YOU LOSE!\n\n";
    }
    hiddenWord = ""; //resetting worrd
}

void guessChar()
{
    char currGuess;
    cout << "\n\nEnter a letter to guess.\n\n";
    cin >> currGuess;

    for (int j = 0; j < currentWord.length(); j++)
    {
        if (currentWord.at(j) == currGuess)  //if guessed char is same as currentWord char iteration
        {                                      // then replace '_' char with letter
            cout << "\n\nLetter " << currGuess << " is in word!\n\n";
            hiddenWord[j] = currGuess;
            guesses ++; //increment guess counter
        }
    }
    if (currentWord.find(currGuess) == string::npos) //if guessed character isnt in word
    {
        cout << "\n\nLetter " << currGuess << " is not in word!\n\n";
        guesses ++;//increasing counter of guesses, if guessed incorrectly
    }
}

string hideWord()
{
    for (int k = 0; k < currentWord.length(); k++)
    {
        hiddenWord += '_';  //replacing word with underlines
    }
    return hiddenWord;
}

void start()
{
    currentWord = randomizeWord(); //calling function and assigning return value as word
    cout << "\n\nNew word has been selected.\n\n";
    hideWord();
    while (guesses <= 6) //game will play as long as guesses below 6
    {
       cout << "(" << hiddenWord << ")" << endl <<endl;
       string response;
       cout << "\n\n(1)Guess letter\n(2)Guess word\n\n";
       cin >> response;

       if (response == "1")
       {
           cout << "You have " << maxGuess - guesses << " guesses remaining!";
           guessChar(); //calling guess function repeatedly
           if (hiddenWord.find('_') == string::npos) //if '_' isnt found, then player has guessed word by finding all chars
           {
               cout << "You guessed all the letters, you win!";
               hiddenWord = "";
               break;
           }
       }
       else if (response == "2")
       {
           guessWord(); //player wants to guess word
           break;
       }

    }
    if (guesses >= 6 && hiddenWord.find('_') != string::npos) //if guesses hit 6, and '_'s are not all gone, then display this
    {
        cout << "\n\nOut of guesses, GAME OVER!\n\n";
    }
}
void menu()
{
    while (true)
    {
        string userResponse;

        cout << "\n\nPlease enter number response.\n\n(1)Play\n(2)Rules\n(3)Exit\n\n";
        cin >> userResponse;

        if (userResponse == "1")
        {
            start();
        }
        else if (userResponse == "2")
        {
            cout << displayRules();
        }
        else if (userResponse == "3")
        {
            cout << "Goodbye!";
            break;
        }
        else
        {
            cout << "\n\nPlease enter valid input\n\n";
        }
    }
}

int main()
{
    cout << "Welcome to hangman!";

    menu();
    return 0;
}

File: python/Beginner/hangman/hangman.py
#python hangman game
import random

ran_int = random.randint(0,4) #generating a random int for word seed
words = ['elephant', 'table', 'uniform', 'danger', 'invite'] #words that the game will choose from
current_word = words[ran_int] #assigning value to word from array via ran num seed
max_guesses = 6
guess_count = 0   
hidden_word = '' 
already_guessed = []#array to store users past guesses

def guess_word(): #function for guessing the word, only 1 chance
    user_guess = input("Please enter the word you think it is. You only have one try!\n")
    if user_guess == current_word:
        print("Well done! You guessed the word!\n")
    else:
        print("Sorry that was not the word, Game over!\n")

for letters in range(len(current_word)): #length of hidden word
    hidden_word += '_' #creating a hidden word same length as current word

print('\n',hidden_word,"\n")

while guess_count < 6: #while loop until 6 guesses have been made
    if '_' not in hidden_word: #if all chars have bee guessed this will be printed and game will finish
        print("You have guessed all of the letters, well done you win!\n")
        break
    while True: #nested while loop to ensure user cant guess same letter twice
        print("You have ", max_guesses - guess_count, " guesses left!\n")
        guessed_char = input("Please guess a single letter. Or enter 1 to guess the word.\n")
        if guessed_char in already_guessed:
            print("That letter has already been guessed, try another. You have not lost a guess.\n")
        else:
            break
    already_guessed.append(guessed_char)#adding char into an array, so that we can verify if its already been used
    
    if guessed_char in current_word:
        print("Nice! That letter was in the word.\n")
        for letter in range(len(current_word)):   #looping through range of word
            if current_word[letter] == guessed_char:    #if the guessed char is equivalent to curren_word current index  -->        
                hidden_word = hidden_word[:letter] + guessed_char + hidden_word[letter + 1:] #then replace '_' char with original letter  
                print(hidden_word) #strings are immutable in python!! remember this!!!!!           
    elif guessed_char == "1":
        guess_word()
        break             
    else:
        print("That was not in the word!\n")
    guess_count +=1

if guess_count == 6 and '_' in hidden_word:#if guesses reach 6 and the full word hasnt veen guessed then game over
    print("Game over, out of guesses!\n")
elif guess_count == 6 and '-' not in hidden_word:#else if they reach 6 guesses and it has been guessed
    print("You guessed all the letters within 6 guesses well done!")

